### PR TITLE
Restore DB using native PDO

### PR DESCRIPTION
**Pull request by mvorisek**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Restore DB using native PHP/PDO. It is faster and requires no exec function. Also dump DB using `ifsnop/mysqldump-php` package if installed manually.
| Type?             | improvement
| Category?         | TE
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | #20708
| How to test?      | GH CI must pass
| Possible impacts? | no

Spin of #23911

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/23935)

### CODE REVIEW

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
I'm sorry again, but we will not use a dependency we don't require for this project. Feel free to create your own module to use mysqldump if you are not able to use the exec command. I also recommend you to use mysqldump command instead of wrapper with PHP.

**Comment by mvorisek** [comment]
@PierreRambaud this is fix for enviroments with `exec` disabled, dependency is not required, only code can fallback to it if it is installed on the developer side

**Comment by PierreRambaud** [comment]
> @PierreRambaud this is fix for enviroments with `exec` disabled, dependency is not required, only code can fallback to it if it is installed on the developer side

If exec is disabled, use `mysqldump` CLI command, or create your own module to do it.

**Comment by mvorisek** [comment]
No, this fixes the CLI installation which use the dumper class.

**Comment by PierreRambaud** [comment]
> No, this fixes the CLI installation which use the dumper class.

Afaik, it is only used for testing purposes. If you are not able to use exec on your dev platform, I think you will have a bad day :thinking:

**Comment by mvorisek** [comment]
> 
> 
> > No, this fixes the CLI installation which use the dumper class.
> 
> Afaik, it is only used for testing purposes. If you are not able to use exec on your dev platform, I think you will have a bad day 🤔

Yes, we have strict environment to avoid these hacky/unmanaged things. I done my best to fix this cleanly, if you think even this solution is not good, I will revert the last commit. Please advise.

**Comment by mvorisek** [comment]
This PR is highly welcomed as `mysqldump` is no longer included by default in `ubuntu-latest` images, see https://github.com/PrestaShop/PrestaShop/pull/23954#issuecomment-824621868

**Comment by PierreRambaud** [comment]
> This PR is highly welcomed as `mysqldump` is no longer included by default in `ubuntu-latest` images, see [#23954 (comment)](https://github.com/PrestaShop/PrestaShop/pull/23954#issuecomment-824621868)

The mysqldump command can be installed, this PR still added code that is not inserted in the vendor directory. I'm still against merging it.

**Comment by mvorisek** [comment]
A few lines that can utilize native, but sadly GPL lib, but this library is not distributed nor included with composer.json. In CI or for development purposes, we can use any tool we like, as long as it is free and non bundled. Currently, DB dump/restore between tests is slow and with this PR it can be improved.

**Comment by PierreRambaud** [comment]
> A few lines that can utilize native, but sadly GPL lib, but this library is not distributed nor included with composer.json. In CI or for development purposes, we can use any tool we like, as long as it is free and non bundled. Currently, DB dump/restore between tests is slow and with this PR it can be improved.

PrestaShop is an application and require vendor directory. PrestaShop software is not designed to work as dependency, you've to keep it in mind.

**Comment by mvorisek** [comment]
@PierreRambaud I know. This PR adds a support for the developer to use native/debugable/faster solution. If he wants to use it for the development purposes, he is responsible to install the dependency on it's own.

**Comment by matks** [comment]
Hello @mvorisek 

I agree with @PierreRambaud 

> This PR adds a support for the developer to use native/debugable/faster solution. If he wants to use it for the development purposes, he is responsible to install the dependency on it's own.

If we accept this PR, we also accept the burden of maintainance that comes with it.
If we accept this PR, we also say publicly "The DatabaseDump php class can work with either `mysqldump`or `ifsnop/mysqldump-php`" and we accept that, in the future years, we will do the necessary maintainance and update work to keep it working.

I am sorry but we do not want to accept the burden of maintainance that comes with this PR. We think it will only solve the issue for a very small number of people, and that available workarounds exist and are simple to setup.

This is open source. If you disagree with the project direction you can fork it and introduce the changes you want. But the repository https://github.com/PrestaShop/PrestaShop/ is maintained by us, not you. So we decide what we include in the maintainance scope and what we dont include.

We thank you for your many contributions and I am sorry you spent time to build this PR but we dont want to merge it as we disagree with both the idea and the maintainance cost.

**Comment by mvorisek** [comment]
@matks This PR has 50 lines and does solve two purposes. Dump and restore. Restore is now done using PDO which is completely valid. Dump - the maintanence costs you mentioned - takes about 10 lines.

Please let me know, if the 10 lines with native/debugable dump solution is really that much problematic. If yes, I will remove and keep only the restore with PDO. Please also note, the way the original dump is done, no error messages etc. are supressed from the user.

**Comment by matks** [comment]
Hello @mvorisek 

> @matks This PR has 50 lines and does solve two purposes. Dump and restore. Restore is now done using PDO which is completely valid. Dump - the maintanence costs you mentioned - takes about 10 lines.

I confirm that the 10 lines cannot be merged. Even if it was 1 line, the issue is not the number of line, is the introduction of `ifsnop/mysqldump-php`.

<hr>

About using PDO to restore database rather than mysqldump: this could be a good idea 👍 .

I have no idea what is faster, load data using `exec('mysql DATABASE < DUMP 2> /dev/null';` or `$PDO->exec(...');` .

Do you have benchmarks or analysis I could reproduce so I can verify it is indeed faster?

Also do you know what is inside `$this->dumpFile` ? Sometimes `$pdo->exec(...)` requires [specific settings](https://stackoverflow.com/a/32354533) to run

**Comment by PierreRambaud** [comment]
> Hello @mvorisek
> 
> > @matks This PR has 50 lines and does solve two purposes. Dump and restore. Restore is now done using PDO which is completely valid. Dump - the maintanence costs you mentioned - takes about 10 lines.
> 
> I confirm that the 10 lines cannot be merged. Even if it was 1 line, the issue is not the number of line, is the introduction of `ifsnop/mysqldump-php`.
> 
> About using PDO to restore database rather than mysqldump: this could be a good idea  .
> 
> I have no idea what is faster, load data using `exec('mysql DATABASE < DUMP 2> /dev/null';` or `$PDO->exec(...');` .
> 
> Do you have benchmarks or analysis I could reproduce so I can verify it is indeed faster?
> 
> Also do you know what is inside `$this->dumpFile` ? Sometimes `$pdo->exec(...)` requires [specific settings](https://stackoverflow.com/a/32354533) to run

DatabaseDump is only used for testing, we clearly don't care about using pdo or mysqldump as we control the CI and we can install mysqldump tool, like we now have to include mysql-default-client because client and server are also in separated packages.

**Comment by mvorisek** [comment]
I have reverted the dump using the GPL lib even if the dependency was optional.

The advantage using native PDO is that errors are handled and `createPdo` method there can be a good start for someone who wants to help with a native dump solution that satisfy the licenses.

File: src/PrestaShopBundle/Install/DatabaseDump.php
Comment (PierreRambaud) [inline src/PrestaShopBundle/Install/DatabaseDump.php:71]: Duplicate from Db class, I'm sorry, but your PR brings no value here. I see no use case where it's needed to change the default behaviour

Comment (mvorisek) [reply src/PrestaShopBundle/Install/DatabaseDump.php:71]: If it can be deduped, please submit a code sudgestion.

Comment (PierreRambaud) [reply src/PrestaShopBundle/Install/DatabaseDump.php:71]: It's up to you to convince us to merge your PR, more if this PR brings no value :thinking:

Comment (mvorisek) [reply src/PrestaShopBundle/Install/DatabaseDump.php:71]: Pay me for that :)

I can stay confident that handling (and showing as well, which is not the case with exec) possible DB errors is enough value to merge this in.

Comment (PierreRambaud) [reply src/PrestaShopBundle/Install/DatabaseDump.php:71]: > Pay me for that :)

Would you pay me to review and merge your pull request?

Comment (mvorisek) [reply src/PrestaShopBundle/Install/DatabaseDump.php:71]: never and thanks for the PS t-shirt 8-)

**Comment by matks** [comment]
Hi everybody,

I close this PR because I believe the tone is slowly becoming too heated.

@mvorisek I am sorry but I think it's better not to continue with this PR. @PierreRambaud  is a maintainer of this project and is not convinced at all whether introducing this change into the project is a good thing. @PierreRambaud  has been trusted to become a maintainer of PrestaShop and as such we need to trust his input concerning what benefits the project and what does not benefit.

I dont think there is value arguing about this little snippet, especially since this is a dev tool, not used in production, right ?

@mvorisek  if we do not merge your PR it does not mean we do not value your work and your skills 😉 . We juste disagree whether this change benefits the project. It's not an issue: it just means we are aiming for a direction and you are aiming for another direction. There is no "bad" or "good" direction, just choices.

@PierreRambaud regularly 'request changes' my Pull Requests 😆  and that's OK it's just working together and sometimes I overlook something and he signals it to me. I trust him to block my PRs when I do something that is not valuable.

As I said before, when we validate a PR it means we also accept to associate our name to it and to defend its benefits. 
When we validate a PR we become as responsible as you about this change being introduced in the project. In the future, we'll be blamed if the change finally introduces issues. That is why we prefer sometimes not to merge something if we are not fully confident.

I dont think the discussion here is benefitting the project and the initial topic (optimizing a dev tool) might not be worth the time and the energy of the participants. 😉  Let's move on to something else, a lot more interesting and less heated.

**Comment by mvorisek** [comment]
@matks let's do a SWOT analysis together.

Tone - please be very careful before you cut a branch under developers that do not have time to respond to a dummy questions. I value feedback, but be very careful to waste my resources. As you can see, no contributions are comming already :D

Now the analysis.

I am sure that it is obvious that calling a binary for something that we can do natively in php is absurd as even more absurd when no possible errors are handled. This PR fixes it.

Weaknesses. Tell me. This PR works, errors are handled, it is not slower, it does not need any system binary. I do not see any.

**Comment by matks** [comment]
Hi @mvorisek sorry for very late answer.

It's hard to write an answer because I really fail to understand what's going on 😞 . With @PierreRambaud  we tried to explain many times why we are not willing to merge this change but it seems you don't see.

One one hand I would like to clear any misunderstanding and I would like you to understand our opinion on the matter: that this contribution brings very few value AFAIK, for a very unusual usecase, that we see no impact or improvement in changing this part of the code for the project need (you might benefit from this change, but your usecase and the project usecase might be different), that then without much value we dont see why the PR would have to be merged... and that you have others contributions that are lot more valuable (and got merged https://github.com/PrestaShop/PrestaShop/pull/23975 https://github.com/PrestaShop/PrestaShop/pull/23960)... 👍 

On another hand I dont want to write more and more and again and again the same thing I feel like we are going in circles. 😞 

I'm sorry I dont see how to explain better why we are not willing to merge this PR. When I read this github comments again everything has been said. I dont know what else to do.

Once again we thank you for your many contributions and the time you spent on this PR and I am sorry I will not approve it. I think there are other topics you have mentioned / started that are a lot more interesting for you and us 👍 this one has wasted already a lot of both your time and ours, I think it's better to not continue.

Note that on this repository, a Pull Request needs 2 approvals to be mergeable. If @PierreRambaud and I do not think this PR is a good fit, other maintainers might have a different opinion.

On my part I will not get involved anymore on this PR 😞 I dont see what else I can bring

**Comment by mvorisek** [comment]
I am rereading the discussion history and I am quite sad, as I see valuable change in the proposed changes.

**Pull request closed without merging**
